pkg/api/v1/types.go

```diff
@@ -800,6 +800,9 @@ type ClusterSpec struct {
 	// ClusterNetwork contains network settings.
 	ClusterNetwork *kubermaticv1.ClusterNetworkingConfig `json:"clusterNetwork,omitempty"`
 
+	// ExternalCloudProvider enables the automatic migration from in-tree cloud provider to external CCM/CSI
+	ExternalCloudProvider *bool `json:"externalCloudProvider,omitempty"`
```

Comment:
I'm worried that someone might use this to disable the `ExternalCloudProvider` feature when it should be enabled. The external CCM should be enabled by default for all newly created clusters that support the external CCM (there's a function for checking that).

Should we extend validation to validate create requests and reject the request if the provider supports the external CCM, but it's disabled?

Reply:
There is an aspect not clear to me: [`ExternalCloudControllerFeatureSupported`](https://github.com/kubermatic/kubermatic/blob/4e1b8c631a9f09cea31b666859b127d2c13fef56/pkg/resources/cloudcontroller/util.go#L29) automatically enables the `externalCloudProvider` feature if the provider supports it. My question is: should be the user allowed to choose if to enable it or not? If not, this flag appears to be useless to me.

Reply:
> There is an aspect not clear to me: ExternalCloudControllerFeatureSupported automatically enables the externalCloudProvider feature if the provider supports it.

It's enabled by the following code for the new clusters: https://github.com/kubermatic/kubermatic/blob/cc2db9dad1faea9e65a3c5421987cdf050262ab6/pkg/handler/common/cluster.go#L191-L193

> My question is: should be the user allowed to choose if to enable it or not? If not, this flag appears to be useless to me.

For new clusters, they shouldn't be able to choose.

For existing clusters, they should be able to enable the `ExternalCloudProvider` feature if it isn't enabled. That would trigger the CCM/CSI migration. From that standpoint, this field might be useful, but you should check that with the SIG API.

Reply:
As discussed, let's remove this and move the information to the cluster status. Changes (migration) can be done using the POST endpoint (to be implemented).